client/db/bolt/db.go

```diff
@@ -532,24 +559,53 @@ func (db *BoltDB) marketOrdersSince(dexB, baseB, quoteB []byte, n int, since uin
 		timeB := oBkt.Get(updateTimeKey)
 		return bEqual(dexB, oBkt.Get(dexKey)) && bEqual(baseB, oBkt.Get(baseKey)) &&
 			bEqual(quoteB, oBkt.Get(quoteKey)) && bytes.Compare(timeB, sinceB) >= 0
-	})
+	}, true)
 }
 
 // newestOrders returns the n newest orders, filtered with a supplied filter
 // function. Each order's bucket is provided to the filter, and a boolean true
 // return value indicates the order should is eligible to be decoded and
 // returned.
-func (db *BoltDB) newestOrders(n int, filter func([]byte, *bbolt.Bucket) bool) ([]*dexdb.MetaOrder, error) {
+func (db *BoltDB) newestOrders(n int, filter func([]byte, *bbolt.Bucket) bool, includeArchived bool) ([]*dexdb.MetaOrder, error) {
 	var orders []*dexdb.MetaOrder
-	return orders, db.ordersView(func(master *bbolt.Bucket) error {
-		pairs := newestBuckets(master, n, updateTimeKey, filter)
-		for _, pair := range pairs {
-			oBkt := master.Bucket(pair.k)
-			o, err := decodeOrderBucket(pair.k, oBkt)
+	return orders, db.ordersView(func(ob, archivedOB *bbolt.Bucket) error {
+		buckets := []*bbolt.Bucket{ob}
+		if includeArchived {
+			buckets = append(buckets, archivedOB)
+		}
+		idx := newTimeIndexNewest(n)
+		for i, master := range buckets {
+			err := master.ForEach(func(k, _ []byte) error {
+				bkt := master.Bucket(k)
+				stamp := intCoder.Uint64(bkt.Get(updateTimeKey))
+				if filter == nil || filter(k, bkt) {
+					// Append the bucket position to the key
+					// so that we know which bucket it is later.
+					k = append(k, byte(i))
```

Comment:
I see what you're doing in this commit, but modifying the key's `[]byte` that `ForEach` provides makes me super uncomfortable.  The `bbolt.Cursor` docs indicate the key and value byte slices are only valid for the life of the transaction, which is not the problem per se, except that this indicates these are _bolt-internal buffers_ that should only be read or copied.

Also TBH I really liked the simplicity of the previous `newestOrders`, which could stay simple with just a sort at the end.  The sort would involve few swaps if the active orders bucket is first.  Open to other solutions though.

Reply:
> The bbolt.Cursor docs indicate the key and value byte slices are only valid for the life of the transaction, which is not the problem per se, except that this indicates these are bolt-internal buffers that should only be read or copied.

I believe `newestBuckets` was already doing this.

I really just adapted that method, because it looks like `timeIndexNewest` is good for handling our set of `n` orders.

Will push two commits, one that fixes the copy issue, and another that just uses `newestBuckets` and sorts and you can decide which is better.

Reply:
I'm referring to `append(k, byte(i))` changing `k` from `ForEach`.

Reply:
Yeah, it is just reading. I see now.